core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -253,6 +252,9 @@ final class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: Stri
    */
   def delayed: Int = timeoutTimer.size
 
+  /**
+    * Cancel watching on any delayed operations for the given key. Note the operation will not be completed
+    */
```

Comment:
Hmm, it seems that there is still a problem with an infinite timeout in DelayedTxnMarker. The issue is that the operation will be added to DelayedOperationPurgatory.timeoutTimer, but will never be removed, which leads to memory leak.